pkg/utils/utils.go

```diff
@@ -19,3 +28,125 @@ func GetOCMAccessToken() (*string, error) {
+		"(display_name = '%s' or cluster_id = '%s' or external_cluster_id = '%s') and "+
+			"status in ('Reserved', 'Active')",
+		key, key, key,
+	)
+	subsListResponse, err := subsResource.List().
+		Search(subsSearch).
+		Size(1).
+		Send()
+	if err != nil {
+		err = fmt.Errorf("Can't retrieve subscription for key '%s': %v", key, err)
+		return
+	}
+
+	// If there is exactly one matching subscription then return the corresponding cluster:
+	subsTotal := subsListResponse.Total()
+	if subsTotal == 1 {
+		id, ok := subsListResponse.Items().Slice()[0].GetClusterID()
+		if ok {
+			var clusterGetResponse *cmv1.ClusterGetResponse
+			clusterGetResponse, err = clustersResource.Cluster(id).Get().
+				Send()
+			if err != nil {
+				err = fmt.Errorf(
+					"Can't retrieve cluster for key '%s': %v",
+					key, err,
+				)
+				return
+			}
+			cluster = clusterGetResponse.Body()
+			return
+		}
+	}
+
+	// If there are multiple subscriptions that match the cluster then we should report it as
+	// an error:
+	if subsTotal > 1 {
+		err = fmt.Errorf(
+			"There are %d subscriptions with cluster identifier or name '%s'",
+			subsTotal, key,
+		)
+		return
+	}
+
+	// If we are here then no subscription matches the passed key. It may still be possible that
+	// the cluster exists but it is not reporting metrics, so it will not have the external
+	// identifier in the accounts management service. To find those clusters we need to check
+	// directly in the clusters management service.
+	clustersSearch := fmt.Sprintf(
+		"id = '%s' or name = '%s' or external_id = '%s'",
+		key, key, key,
+	)
+	clustersListResponse, err := clustersResource.List().
+		Search(clustersSearch).
+		Size(1).
+		Send()
+	if err != nil {
+		err = fmt.Errorf("Can't retrieve clusters for key '%s': %v", key, err)
+		return
+	}
+
+	// If there is exactly one cluster matching then return it:
+	clustersTotal := clustersListResponse.Total()
+	if clustersTotal == 1 {
+		cluster = clustersListResponse.Items().Slice()[0]
+		return
+	}
+
+	// If there are multiple matching clusters then we should report it as an error:
+	if clustersTotal > 1 {
+		err = fmt.Errorf(
+			"There are %d clusters with identifier or name '%s'",
+			clustersTotal, key,
+		)
+		return
+	}
+
+	// If we are here then there are no subscriptions or clusters matching the passed key:
+	err = fmt.Errorf(
+		"There are no subscriptions or clusters with identifier or name '%s'",
+		key,
+	)
+	return
+}
+
+func GetClusterLimitedSupportReasons(connection *sdk.Connection, clusterID string) ([]*lmtSprReasonItem, error) {
+
+	limitedSupportReasons, err := connection.ClustersMgmt().V1().
+		Clusters().
+		Cluster(clusterID).
+		LimitedSupportReasons().
+		List().
+		Send()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get limited Support Reasons: %s", err)
+	}
+
+	lmtReason := limitedSupportReasons.Items()
+
+	var clusterLmtSprReasons []*lmtSprReasonItem
+
+	lmtReason.Each(func(lmtSprReason *cmv1.LimitedSupportReason) bool {
```

Comment:
What is this each doing?

Reply:
Hey @fahlmant this is because we get a list when we get the limited support reasons. This is because a cluster can have many reasons why it is on limited support. Thus, for each of those reasons (1 or many) on that limited support list, we create a reason object and append them to the slice we will return. We do this Right here instead of let's say just return the full list to the code calling this function because we also want to get only the pertinent properties (the ones we care about) for the limited support reasons and because that way if we need this code at some point we don't need to parse that list, we just get the slice of reasons we need and job done. And if we want more or less at some point we can modify this struct and the .each part of the code.

So In summary, we looping through each limited support reason and extracting only the info we need.